### PR TITLE
Enable Rust to use the EHCont security feature of Windows

### PR DESCRIPTION
**Pull request by sivadeilra**
In the future Windows will enable Control-flow Enforcement Technology (CET aka Shadow Stacks). To protect the path where the context is updated during exception handling, the binary is required to enumerate valid unwind entrypoints in a dedicated section which is validated when the context is being set during exception handling.

The required support for EHCONT Guard has already been merged into LLVM, long ago. This change simply adds the Rust codegen option to enable it.

Relevant LLVM change: https://reviews.llvm.org/D40223

This also adds a new `ehcont-guard` option to the bootstrap config which enables EHCont Guard when building std.

We at Microsoft have been using this feature for a significant period of time; we are confident that the LLVM feature, when enabled, generates well-formed code.

We currently enable EHCONT using a codegen feature, but I'm certainly open to refactoring this to be a target feature instead, or to use any appropriate mechanism to enable it.

### CODE REVIEW

**Comment by TaKO8Ki** [comment]
Reroll r? compiler

File: compiler/rustc_session/src/options.rs [hunk 1]
```diff
@@ -1387,6 +1387,8 @@ options! {
         "allow the linker to link its default libraries (default: no)"),
     dlltool: Option<PathBuf> = (None, parse_opt_pathbuf, [UNTRACKED],
         "import library generation tool (ignored except when targeting windows-gnu)"),
+    ehcont_guard: bool = (false, parse_bool, [TRACKED],
+        "generate Windows EHCont Guard tables"),
```
Comment (Nilstrieb) [inline compiler/rustc_session/src/options.rs:1391]: This is an insta-stable option, and would need an FCP to be added. Can you instead add it as an unstable option first?

**Review by Noratrieb — COMMENTED** [review]
@wesleywiser @ChrisDenton you may have context and opinions on how this should best be implemented

File: compiler/rustc_session/src/options.rs (hunk 1 above)
Comment (sivadeilra) [reply compiler/rustc_session/src/options.rs:1391]: Certainly.  Would you prefer this as an unstable codegen option, or an unstable target feature?

Comment (Nilstrieb) [reply compiler/rustc_session/src/options.rs:1391]: I'm not sure what's better, I'll let the others decide.

Comment (wesleywiser) [reply compiler/rustc_session/src/options.rs:1391]: The normal path for adding codegen options like this is first as a `-Z` flag and then FCP the `-C` stabilization. 

@sivadeilra if you don't mind changing this to be an unstable option currently, I'm happy to r+ this!

File: src/bootstrap/src/core/builder.rs [hunk 2]
```diff
@@ -1963,6 +1963,12 @@ impl<'a> Builder<'a> {
             rustflags.arg("-Ccontrol-flow-guard");
         }
 
+        // Same for EHCont Guard (this is not combined with the previous if-statement to make
+        // merges with upstream easier).
```
Comment (wesleywiser) [inline src/bootstrap/src/core/builder.rs:1967]: nit: Comment should probably be adjusted 🙂

Comment (sivadeilra) [reply src/bootstrap/src/core/builder.rs]: Updated comment.  Thanks!

File: compiler/rustc_session/src/options.rs (hunk 1 above)
Comment (sivadeilra) [reply compiler/rustc_session/src/options.rs]: Converted to `-Z ehcont-guard`

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
Thanks @sivadeilra!

@bors r+ rollup

**Pull request merged**